### DISCUSSION TITLE
Sign in methods should all support adding user metadata

### DISCUSSION THREAD

**Discussion by elyobo** [supabase/supabase#29075] [Feature Requests]
`supabase.auth.signUp` supports setting user metadata in the auth options but the sign in methods like `supabase.auth.signInWithPassword` or `supabase.auth.signInWithIdToken` do not. It would be useful to be able to set information as part of that sign in flow.

My use case is that I sign in a user with Google, then call an internal API via an edge function which performs additional validation and adds data to the user app metadata that is then processed in a custom auth hook for the claims. I'd like to clear the user's custom metadata when initially signing in so that the custom claim processing does not apply to any old metadata which is still present.

**Comment by v-jagbo** [comment]
Thank you for bringing up this important feature request regarding the inclusion of user metadata in sign-in methods like supabase.auth.signInWithPassword and supabase.auth.signInWithIdToken. I completely understand how beneficial it would be to have the ability to set user metadata during the sign-in flow, especially in scenarios like yours where you require additional validation and custom claim processing.

Currently, as you noted, the supabase.auth.signUp method supports setting user metadata, but the sign-in methods do not. This limitation can pose challenges for developers who want to manage user metadata seamlessly throughout the authentication process.

One potential solution could be to introduce an optional parameter in the sign-in methods that allows developers to pass user metadata alongside the authentication request. This would enable you to clear or update the user's custom metadata during sign-in, ensuring that subsequent claim processing is based on the most relevant information.

I encourage others in the community to weigh in on this suggestion as well. The more feedback we can provide, the more likely it is that this feature will be considered for future updates.

Thank you for advocating for an enhancement that can improve the Supabase authentication experience.

**Reply by marcospgp** [reply]
chatgpt reply

**Comment by v-jagbo** [comment]
This is a possible solution for the User Metadata Management in Sign-In Flow which can be adopted by Supabase. 

1. Enhance Sign-In Method Parameters:

Modify the supabase.auth.signInWithPassword and supabase.auth.signInWithIdToken methods to accept an optional metadata parameter for setting user metadata directly during the sign-in process.

![image](https://github.com/user-attachments/assets/27bd8d98-5741-4e3b-ab27-ae950e1f0651)

2. Introduce a clearMetadata Option:

Add an option named clearMetadata that allows developers to clear existing user metadata upon signing in. This ensures that any leftover metadata from previous sessions does not interfere with the current user's claims.

![image](https://github.com/user-attachments/assets/c5797037-285f-4b5b-bacf-4bb06b84b1e3)

3. Implement Edge Function for Additional Metadata Handling:

After a successful sign-in, invoke an internal API via an edge function to perform additional validation and update the user app metadata. This function will process any necessary data and ensure that the metadata remains relevant.

![image](https://github.com/user-attachments/assets/a2ecbb2d-bcc1-4502-a8fd-7097a6e07ca7)

4. Custom Auth Hook for Claims Processing:

Utilize a custom auth hook to process the updated metadata for claims. This ensures that only fresh, relevant metadata is considered when determining user roles or permissions.

**Reply by elyobo** [reply]
Appreciate the feedback, but this all reads extremely chatgpt-ish.

Additional processing e.g. in custom auth hooks is outside the scope of this feature request; it's what I specifically want to do with it but any implementation should be generic enough to support that and other use cases.

The ability to set metadata, as is possible in the sign up call, is probably sufficient. This is also consistent with `supabase.auth.signUp()` and also `supabase.auth.updateUser()`. The docs are not explicit, but I assume that the provided `data` parameter fully overwrites the current metadata rather than merging with it, so simply providing `{}` would remove all existing metadata if this was desired.

**Reply by GaryAustin1** [reply]
The JSON data should be merged, not overwritten.

**Reply by elyobo** [reply]
The docs aren't clear in this regard and I haven't tested it, but if that's the case (and that seems safer!) then a way to remove fields is probably desirable too. Presumably `undefined` will be stripped out by JSON serialisation, so you could send `NULL` to overwrite unwanted values but that would leave keys behind.

That said, the existing methods don't have explicit "remove this value" functionality then so this feature wouldn't need to add it; it would make sense to be consistent with the two existing implementations. If the removal of specific keys is required then it should be added to those as well.

**Reply by GaryAustin1** [reply]
I've seen ways to remove the data with either null or empty {} with the JSON "keys".  But can't look tonight to point you to them.
I know for sure it mergers as I've used it.

**Reply by elyobo** [reply]
Yeah, nulling should work, but that isn't my feature request here (just added as context by the other commenter, who appears to be channeling chatgpt) - however you can only set metadata via `signUp` or `updateUser`, not on the other sign in methods and my feature request is that you should be able to set metadata when signing in a user across _all_ sign in methods.

**Reply by GaryAustin1** [reply]
Yes, I was only commenting on your assertion about updateUser...
It is a constant issue there is nothing built in to do the same for OAuth in particular.

**Comment by hubertoschusch** [comment]
Why is this still not implemented?

**Reply by nandorcsupor** [reply]
Good question

**Comment by pierroo** [comment]
Still not implemented, but is it even considered / in the roadmap?
Did you guys find any update about it?

Having to hack around it feels ugly at the moment.

**Reply by elyobo** [reply]
Have not seen any suggestion that it's under consideration. You just have to sign users in and then set metadata, which is annoying because it should be a single atomic operation. I set some global app state to say that I'm signing the user in, sign them in, set metadata, and clear the "signing the goddam user in" state so that it seems atomic to the user at least.
